File: src/dbscanrn.cpp
#include <iostream>
#include <xtensor/xarray.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xview.hpp>
#include <xtensor/xarray.hpp>
#include <xtensor/xfixed.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xtensor.hpp>
#include <xtensor/xadapt.hpp>
#include <xtensor/xarray.hpp>
#include <xtensor/xaxis_slice_iterator.hpp>
#include <xtensor/xaxis_iterator.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xarray.hpp>
#include <xtensor/xindex_view.hpp>
#include <xtensor/xchunked_array.hpp>
#include <xtensor/xfixed.hpp>
#include <cstddef>
#include <vector>
#include <xtensor/xarray.hpp>
#include <xtensor/xadapt.hpp>
#include <istream>
#include <fstream>
#include <iostream>
#include <xtensor/xarray.hpp>
#include <xtensor/xcsv.hpp>
#include <iostream>
#include <fstream>
#include <functional>
#include <cstdio>
#include <xtensor/xcsv.hpp>
#include <iostream>
#include <chrono>
#include <unistd.h>
#include <fstream>
#include <nlohmann/json.hpp>
#include <chrono>
#include <list>
#include <xtensor/xsort.hpp>
#include <string>
#include <regex>

using namespace xt;
using namespace std;
using namespace std::chrono;
using json = nlohmann::json;

double current_time()
{
    auto current_time = std::chrono::system_clock::now();
    auto duration_in_seconds = std::chrono::duration<double>(current_time.time_since_epoch());
    double num_seconds = duration_in_seconds.count();
    return num_seconds;
}

float cosine_dissimilarity(const xarray<double> &a, const xarray<double> &b)
{
    auto numerator = sum(a * b);
    auto denominator = sqrt(sum(pow(a, 2))) * sqrt(sum(pow(b, 2)));
    auto result = numerator / denominator;
    return 1 - result(0);
}

auto normalize(const xarray<double> &a)
{
    xarray<double> result = a / sqrt(sum(pow(a, 2)));
    return result;
}

xarray<double> CLUSTER;
xarray<double> STATE;

int main()
{

    string path;
    double epsilon;
    int minPts;
    string out_path;
    string log_out;

    // read configs
    std::ifstream file("configs/dbscan.json");
    json config = json::parse(file);

    for (auto conf : config)
    {
        if (not conf["disable"].get<bool>())
        {
            std::cout << conf["name"].get<std::string>() << "\n";
            auto path = conf["path"].get<std::string>();
            auto k = conf["params_dbscanrn"]["k"].get<int>();
            auto out_path = conf["out_path"].get<std::string>();
            auto log_out = conf["log_out"].get<std::string>();

            out_path = regex_replace(out_path, regex("algorithm"), "dbscanrn_cpp");
            out_path = regex_replace(out_path, regex("similarity"), "cosine_dissimilarity");
            log_out = regex_replace(log_out, regex("algorithm"), "dbscanrn_cpp");
            log_out = regex_replace(log_out, regex("similarity"), "cosine_dissimilarity");

            std::cout << out_path << "\n";
            std::cout << log_out << "\n";

            // load data
            ofstream outfile;
            outfile.open(log_out);

            auto start = chrono::high_resolution_clock::now();

            ifstream input_file;
            input_file.open(path);
            auto data = xt::load_csv<double>(input_file);
            input_file.close();

            auto end = chrono::high_resolution_clock::now();
            duration<double> elapsed = (end - start);
            outfile << std::setprecision(16) << current_time() << ",reading_data,"
                    << "," << elapsed.count() << "," << endl;

            data.reshape(data.shape());
            xarray<double> X = view(data, all(), range(0, -1));

            // init
            void clustering_algorithm(const xarray<float> &X_original, int k, std::ofstream &outfile);

            // run algorithm
            cout << "Processing ..."
                 << "\n";
            clustering_algorithm(X, k, outfile);
            start = chrono::high_resolution_clock::now();

            // save output
            ofstream out_file;
            out_file.open(out_path);
            dump_csv(out_file, xt::stack(xtuple(CLUSTER, STATE), 1));
            out_file.close();

            elapsed = (chrono::high_resolution_clock::now() - start);
            outfile << std::setprecision(16) << current_time() << ",writing_data,"
                    << "," << elapsed.count() << ',' << endl;
            outfile.close();
            std::cout << "Done."
                      << "\n";
        }
    }
}

void clustering_algorithm(const xarray<float> &X_original, int k, std::ofstream &outfile)
{
    outfile << std::setprecision(16) << current_time() << ",start log,,," << endl;

    // normalization
    auto start = chrono::high_resolution_clock::now();
    xarray<float> X = normalize(X_original);
    auto end = chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    outfile << std::setprecision(16) << current_time() << ",normalization_time,"
            << "," << elapsed.count() << "," << endl;

    // each data point can be in one of 3 stages
    int NOT_VISITED = -1; // not visited point
    int VISITED = 0;      // non-core point
    int CLUSTERED = 1;    // core point

    // initial setup
    int n = X.shape()[0];
    xarray<double> cluster = ones<double>({n}) * -1;        // cluster register
    xarray<double> state = ones<double>({n}) * NOT_VISITED; // state register
    int cluster_id = 1;
    auto all_point_indices = arange(n);
    map<int, vector<int>> point_rnn;
    map<int, xarray<int>> point_knn;

    auto get_knn = [&](int current_index, xarray<int> neighbor_indices, int k)
    {
        xarray<float> neighbor_similaritys = {0};
        for (auto &neighbor_index : neighbor_indices)
        {

            outfile << std::setprecision(16) << current_time() << ",similarity_calculation," << current_index << ",1," << endl;

            auto similarity = cosine_dissimilarity(
                row(X, neighbor_index), row(X, current_index));
            neighbor_similaritys = concatenate(
                xtuple(neighbor_similaritys, xarray<float>{similarity}));
        };
        xarray<int> sort_indices = argsort(view(neighbor_similaritys, range(1, all())));
        xarray<int> neighbor_indices_sorted = index_view(neighbor_indices, sort_indices);
        point_knn[(int)current_index] = view(neighbor_indices_sorted, range(0, k));
    };

    auto get_pointwise_rnn = [&](int current_index)
    {
        vector<int> rnn;
        for (auto &neighbor_index : point_knn[(int)current_index])
        {
            if ((point_knn[(int)neighbor_index], (int)current_index))
            {
                rnn.push_back((int)neighbor_index);
            };
        };

        point_rnn[(int)current_index] = rnn;
    };

    auto get_rnn = [&]()
    {
        auto start = chrono::high_resolution_clock::now();
        for (int current_index = 0; current_index < n; current_index++)
        {
            // remove current index from neighbors
            xarray<int> neighbor_indices = col(
                from_indices(argwhere(not_equal(all_point_indices, current_index))), 0);
            get_knn((int)current_index, neighbor_indices, k);
        };
        auto end = chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = end - start;

        outfile << std::setprecision(16) << current_time() << ",knn_time,"
                << "," << elapsed.count() << ',' << endl;

        start = chrono::high_resolution_clock::now();
        for (int current_index = 0; current_index < n; current_index++)
        {
            get_pointwise_rnn((int)current_index);

            outfile << std::setprecision(16) << current_time() << ",|rnn_neighbors|," << current_index << "," << (int)point_rnn[(int)current_index].size() << "," << endl;

            outfile << std::setprecision(16) << current_time() << ",rnn_neighbors_id," << current_index << ",,";
            for (auto &i : point_rnn[(int)current_index])
            {
                outfile << i << ";";
            }
            outfile << endl;

            outfile << std::setprecision(16) << current_time() << ",knn_neighbors_id," << current_index << ",,";
            for (auto &i : point_knn[(int)current_index])
            {
                outfile << i << ";";
            }
            outfile << endl;
            outfile << std::setprecision(16) << current_time() << ",|knn_neighbors|," << current_index << "," << (int)point_knn[(int)current_index].shape(0) << "," << endl;
        };
        end = chrono::high_resolution_clock::now();
        elapsed = end - start;

        outfile << std::setprecision(16) << current_time() << ",rnn_time,"
                << "," << elapsed.count() << ',' << endl;
    };

    // extend cluster
    function<void(int)> search = [&](int current_index) -> void
    {
        if (point_rnn[(int)current_index].size() < k)
        {
            state[(int)current_index] = VISITED;
        }
        else
        {
            state[(int)current_index] = CLUSTERED;
            cluster[(int)current_index] = cluster_id;
            for (auto &neighbor_index : point_rnn[(int)current_index])
            {
                if (state[(int)neighbor_index] == NOT_VISITED)
                {
                    search(neighbor_index);
                };
                state[(int)neighbor_index] = CLUSTERED;
                cluster[(int)neighbor_index] = cluster_id;
            };
        };
    };

    get_rnn();

    // visit all points
    while (any(equal(state, NOT_VISITED)))
    {
        xarray<int> idx = from_indices(argwhere(equal(state, NOT_VISITED)));
        search(idx(0, 0));
        cluster_id++;
    }

    // clusterize all outlier points to nearest cluster
    while (any(equal(state, VISITED)))
    {
        xarray<int> idx_not_clustered = from_indices(argwhere(equal(state, VISITED)));
        xarray<int> idx_clustered = from_indices(argwhere(equal(state, CLUSTERED)));
        get_knn((int)idx_not_clustered(0, 0), idx_clustered, 1);
        auto closest_clustered_idx = point_knn[(int)idx_not_clustered(0, 0)](0);
        cluster[(int)idx_not_clustered(0, 0)] = cluster[(int)closest_clustered_idx];
        state[(int)idx_not_clustered(0, 0)] = CLUSTERED;
    }
    outfile << std::setprecision(16) << current_time() << ",stop log,,," << endl;

    CLUSTER = cluster;
    STATE = state;
}

File: src/dbscanrn.py
import logging
import time 
import os
import pandas as pd
from src.metrics import *
from src.datasets import Dataset
import json


def setup_logger(name, log_file, level=logging.INFO):
    """To setup as many loggers as you want"""
    formatter = logging.Formatter(fmt='%(msecs)06f,%(message)s')
    handler = logging.FileHandler(log_file)        
    handler.setFormatter(formatter)

    logger = logging.getLogger(name)
    logger.setLevel(level)
    logger.addHandler(handler)

    return logger

logger = setup_logger('root', 'out.log')


def get_knn(current_index, neighbor_indices, k, similarity, X):
    '''
    current_index - index of given point,
    neighbor_indices - indices of points for searching,
    k - number of neighbors searched for current_index,
    similarity - metric of similarity or distance bewteen two points,
    X - dataset with point coordinates
    '''
    # Remove current point from neighbors
    if current_index in neighbor_indices: neighbor_indices.remove(current_index)
    
    # For each neighbor measure similarity to current point
    neighbor_similaritys = []
    for neighbor_index in neighbor_indices:
        logger.info(f'similarity_calculation, {current_index}, 1,')
        similarity_measure = similarity(X[neighbor_index], X[current_index])
        neighbor_similaritys.append(similarity_measure) 
    
    # Get k-nearest neighbors
    sort_indices = np.argsort(neighbor_similaritys) # sort arguments and return indices
    neighbor_indices = np.array(neighbor_indices) 
    neighbor_indices = neighbor_indices[sort_indices][:k].tolist()
    
    neighbor_indices_str = ';'.join(str(e) for e in neighbor_indices)
    logger.info(f'knn_neighbors_id,{current_index},,{neighbor_indices_str}')
    logger.info(f'|knn_neighbors|,{current_index}, {len(neighbor_indices)},')
    
    return neighbor_indices


def get_pointwise_rnn(point_knn, current_index):
    rnn = []
    for neighbor in point_knn[current_index]:
        if current_index in point_knn[neighbor]:
            rnn.append(neighbor)
    
    rnn_indices_str = ';'.join(str(e) for e in rnn)
    logger.info(f'rnn_neighbors_id,{current_index},,{rnn_indices_str}')
    logger.info(f'|rnn_neighbors|,{current_index}, {len(rnn)},')
    return rnn
    
def get_rnn(point_indices, k, similarity, X):
    '''
    point_indices - indices of points for searching,
    k - number of neighbors searched for current_index,
    similarity - metric of similarity or distance bewteen two points,
    X - dataset with point coordinates
    '''
    point_rnn = {}
    point_knn = {}
    
    timer_start = time.time() 
    for current_index in point_indices:
        knn = get_knn(current_index, point_indices[:], k, similarity, X)
        point_knn[current_index] = knn
    logger.info(f'knn_time,{current_index},{(time.time() - timer_start) * 1000},')
    
    timer_start = time.time()
    for current_index in point_indices:
        rnn = get_pointwise_rnn(point_knn, current_index)
        point_rnn[current_index] = rnn
    logger.info(f'rnn_time,{current_index},{(time.time() - timer_start) * 1000},')
    
    return point_rnn, point_knn

def dbscanrn(X, k, similarity):
    
    logger.info(f'start log,,,')
    
    # each data point can be in one of 3 stages
    NOT_VISITED = -1 # not visited point
    VISTED = 0 # non-core point
    CLUSTERED = 1 # core point
    
    # initial setup
    n = X.shape[0]
    cluster = np.array([-1] * n) # cluster register
    state = np.array([NOT_VISITED] * n) # state register
    cluster_id = 1
    all_point_indices = list(range(len(X))) # inidces of all points
    # calculate RNN_k for all points
    point_rnn, point_knn = get_rnn(all_point_indices, k, similarity, X)
    
    # search for clusters
    def search(current_index, k):
        if len(point_rnn[current_index]) < k:
                state[current_index] = VISTED
        else:
            state[current_index] = CLUSTERED
            cluster[current_index] = cluster_id
            for neighbor_index in point_rnn[current_index]:
                if state[neighbor_index] == NOT_VISITED:
                    search(neighbor_index, k)
                state[neighbor_index] = CLUSTERED
                cluster[neighbor_index] = cluster_id
                    
    # visit all X points
    while NOT_VISITED in state:
        not_visited_ids = np.where(state==NOT_VISITED)[0][0]
        search(not_visited_ids, k)
        cluster_id += 1
    
    # clusterize all outlier data points 
    while VISTED in state:
        not_clustered_ids = np.where(state==VISTED)[0][0]
        clustered_ids = np.where(state==CLUSTERED)[0]
        knn = get_knn(not_clustered_ids, clustered_ids, 1, similarity, X)
        cluster[not_clustered_ids] = cluster[knn[0]]
        state[not_clustered_ids] = CLUSTERED
    logger.info(f'stop log,,,')
    return cluster, state



class DBSCANRN:

    def __init__(self, k, similarity=cosine_dissimilarity, **kwargs):
        self.k = k
        self.similarity = similarity
        self.log_output = 'out.log'
        self.name = 'dbscanrn'

        config_path = './configs/dbscan.json'
        f = open(config_path)
        self.config = json.load(f)

    def run(self, dataset_name):
        
        logger = logging.getLogger('root')
        formatter = logging.Formatter(fmt='%(msecs)06f,%(message)s')    
        handler = logging.FileHandler(self.log_output)
        handler.setFormatter(formatter)
        logger.addHandler(handler)         
        
        logger.info(f'start log,,,')
        
        timer1 = time.time() 
        dataset = Dataset(self.config[dataset_name]['path'], False)
        X, y = dataset.X, dataset.y
        logger.info(f'reading_data,,{(time.time() - timer1)*1000},')

        timer1 = time.time() 
        X = X/np.linalg.norm(X, axis =1, keepdims = True)
        logger.info(f'normalization_time,,{(time.time() - timer1)*1000},')
        
        self.X = X
        result = dbscanrn(self.X, self.k, self.similarity)
        self.y_pred, self.state = result
        
        timer1 = time.time() 
        pd.DataFrame({
            'cluster': self.y_pred,
            'state': self.state
        }).to_csv(f"out/{self.name}.csv", header=None, index=False)
        logger.info(f'writing_data,,{(time.time() - timer1)*1000},')
        
        return self.y_pred
    
    def get_logs(self):
        logs = pd.read_csv(
            self.log_output,
            names=['time [ms]', 'operation', 'point_id', 'value', 'string']
        )
        # logs['time [ms]'] -= logs['time [ms]'].min()
        return logs